tests/e2e/cmux_test.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2023 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// These tests are directly validating etcd connection multiplexing.
+//go:build !cluster_proxy
+
+package e2e
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+	"testing"
+
+	"github.com/prometheus/common/expfmt"
+	"github.com/stretchr/testify/require"
+	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	"go.etcd.io/etcd/api/v3/version"
+	"go.etcd.io/etcd/server/v3/etcdserver/api/etcdhttp"
+	"go.etcd.io/etcd/tests/v3/framework/config"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+	"go.uber.org/zap"
+)
+
+func TestConnectionMultiplexingTLS(t *testing.T) {
+	e2e.BeforeTest(t)
+	cfg := e2e.EtcdProcessClusterConfig{ClusterSize: 1, Client: e2e.ClientConfig{ConnectionType: e2e.ClientTLS}}
+	ctx := context.Background()
+	clus, err := e2e.NewEtcdProcessCluster(ctx, t, e2e.WithConfig(&cfg))
+	require.NoError(t, err)
+	defer clus.Close()
+
+	testConnectionMultiplexing(t, ctx, clus.Cfg.Logger, clus.EndpointsV3()[0], true)
+}
+
+func TestConnectionMultiplexingNonTLS(t *testing.T) {
+	e2e.BeforeTest(t)
+	cfg := e2e.EtcdProcessClusterConfig{ClusterSize: 1, Client: e2e.ClientConfig{ConnectionType: e2e.ClientNonTLS}}
+	ctx := context.Background()
+	clus, err := e2e.NewEtcdProcessCluster(ctx, t, e2e.WithConfig(&cfg))
+	require.NoError(t, err)
+	defer clus.Close()
+
+	testConnectionMultiplexing(t, ctx, clus.Cfg.Logger, clus.EndpointsV3()[0], false)
+}
+
+func TestConnectionMultiplexingTlsAndNonTLS(t *testing.T) {
+	e2e.BeforeTest(t)
+	cfg := e2e.EtcdProcessClusterConfig{ClusterSize: 1, Client: e2e.ClientConfig{ConnectionType: e2e.ClientTLSAndNonTLS}}
+	ctx := context.Background()
+	clus, err := e2e.NewEtcdProcessCluster(ctx, t, e2e.WithConfig(&cfg))
+	require.NoError(t, err)
+	defer clus.Close()
+
+	for _, isTLS := range []bool{true, false} {
+		name := "ClientNoTLS"
+		if isTLS {
+			name = "ClientTLS"
+		}
+		t.Run(name, func(t *testing.T) {
+			testConnectionMultiplexing(t, ctx, clus.Cfg.Logger, clus.EndpointsV3()[0], isTLS)
+		})
+	}
+}
```

Comment:
There are too much duplicated code. Suggest to refactor the test cases using table-driven sub test to remove all the duplication.